src/aum_line_search.cpp

```diff
@@ -1,180 +1,404 @@
+     &M, &id_from_rank, &rank_from_id, slope_from_id, lineCount);
+  int last_line = 0;
+  double last_thresh = intercept_from_id[0];
+  for(int line_i=1; line_i<=lineCount; line_i++){
+    double thresh;
+    if(line_i==lineCount){
+      thresh = INFINITY;
+    }else{
+      thresh = intercept_from_id[line_i];
     }
-    double aum = initialAum;
-
-    for (int iterations = 1; iterations < maxIterations && !intersections.empty(); iterations++) {
-        auto intersection = *intersections.begin();
-        // swap the indices of the lines that make this intersection point
-        // the names of these variables look backwards because they get swapped
-        int lineIndexLowAfterIntersect = backwardsIndices[intersection.lineLowBeforeIntersect];
-        int lineIndexHighAfterIntersect = backwardsIndices[intersection.lineHighBeforeIntersect];
-        swap(backwardsIndices[intersection.lineLowBeforeIntersect], backwardsIndices[intersection.lineHighBeforeIntersect]);
-        swap(sortedIndices[lineIndexHighAfterIntersect], sortedIndices[lineIndexLowAfterIntersect]);
-
-        // indices of the next lines we want to find intersections for
-        int higherLineIndex = lineIndexHighAfterIntersect + 1;
-        int lowerLineIndex = lineIndexLowAfterIntersect - 1;
-
+    if(last_thresh < thresh){
+      total_auc.value += total_auc.get_auc(last_line, line_i);
+      last_line = line_i;
+      last_thresh = thresh;
+    }
+  }
+  aucAtStepVec[0] = total_auc.value;
+  total_auc.zero_update_all();
+  // AUM at step size 0
+  aumVec[0] = aum;
+  aumSlopeAfterStepVec[0] = total_auc.aum_slope;
+  stepSizeVec[0] = 0.0;
+  aucAfterStepVec[0] = total_auc.value;
+  intersectionCountVec[0] = 0;
+  intervalCountVec[0] = 0;
+  qSizeVec[0]=queue.step_ThreshIntervals_map.size();
+  for//iterations/step sizes
+    (int iteration = 1; 
+     iteration < maxIterations && !queue.step_ThreshIntervals_map.empty(); 
+     iteration++){
+    auto TI_it = queue.step_ThreshIntervals_map.begin();
+    double stepSize = TI_it->first;
+    aum += total_auc.aum_slope * (stepSize - lastStepSize);
+    stepSizeVec[iteration] = stepSize;
+    aumVec[iteration] = aum;
+    ThreshIntervals TI = TI_it->second;
+    TI.set_intervals_ranks();
+    double more_auc_at_step = total_auc.action(&TI, -1.0);
+    double auc_after_remove = total_auc.value;
+    intersectionCountVec[iteration] = TI.thresh_intervals_map.size();
+    intervalCountVec[iteration] = 0;
+    for//intersection points. (tie-breaking type 1)
+      (auto intervals_it = TI.thresh_intervals_map.begin(); 
+       intervals_it != TI.thresh_intervals_map.end();
+       intervals_it++){
+      double FPhi_tot=0, FPlo_tot=0, FNhi_tot=0, FNlo_tot=0;
+      intervalCountVec[iteration] += intervals_it->second.n_intervals;
+      int lowest_rank = intervals_it->second.low_rank;
+      int highest_rank = intervals_it->second.high_rank;
+      for//adjacent lines in an intersection point. (tie-breaking type 2)
```

Comment:
A comment at the top of the function / file explaining the two different kinds of tie breaking would be nice

Reply:
ok